connections/hole-punching.md

```diff
@@ -0,0 +1,257 @@
+public IP addresses as well as their port mapping. Examples of STUN-like
+protocols are [STUN] (hence the name) and [AutoNAT].
+
+#### Coordination
+
+Coordination protocols run on top of some relay protocol enabling two hosts to
+coordinate the hole punching process. Examples of coordination protocols are the
+[Session Description Protocol (SDP)][SDP] and [Direct Connection Upgrade through
+Relay][DCUTR].
+
+## Vision
+
+When it comes to establishing direct connections one needs to differentiate by
+platform type (browser / non-browser) as as well as routability (public /
+private). The below describes on a high level how to establish direct
+connections across the various permutations of the two dimensions.
+
+### Public Non-Browser (A) to Public Non-Browser (B)
+
+This is the ideal case, where both hosts can be directly dialed, thus there
+being no need for hole punching. The two hosts should use either TCP or QUIC to
+communicate directly.
+
+### Private Non-Browser (A) to Public Non-Browser (B)
+
+Given that B is publicly routable, A can establish a direct TCP or QUIC
+connection to B.
+
+### Private Browser (A) to Public Non-Browser (B)
+
+Given that B is publicly routable A can establish a direct connection, though
+multiple restrictions are imposed by the browser platform type of A:
+
+- The only transport protocol allowing direct connections on browser platforms
+  is WebSockets. Thus B has to support the WebSocket transport protocol.
+
+- Browser platforms do not allow insecure WebSocket connections, thus B has to
+  have a valid TLS certificate to offer secure Websocket communication.
+
+In cases where B does not fulfill both of these requirements, additional steps
+as detailed in [#private-browser-a-to-private-non-browser-b] are necessary.
+
+### Public or Private Non-Browser (A) to Private Non-Browser (B)
+
+In order to establish a direct connection from A to B, one needs to utilize some
+means of firewall / NAT hole punching. Given the permissiveness of non-browser
+platforms there are many ways to achieve this goal, with the below being one of
+them.
+
+- B uses the [AutoNat] protocol to determine whether it is running behind
+  firewalls and/or NATs and in the latter case what type of NATs.
+  
+  Note: Hole punching fails if either A or B runs behind a [symetric
+  NAT][symetric-nat].
+
+- B establishes a TCP or QUIC connections to one or more relay servers and
+  listens for incoming connection requests via the [circuit relay v2
+  protocol][circuit-relay-v2]. Once B established a reservation at one or more
+  relay servers, B can advertise its relayed addresses (e.g.
+  `/ip4/.../tcp/.../p2p/QmRelay`) via some external mechanism.
+
+  Note on decentralization: Given the low resource requirements of the [circuit
+  relay v2 protocol][circuit-relay-v2] any public host can serve as a relay
+  server.
+
+- A discovers the relayed address of B through some external mechanism. Given
```

Comment:
Maybe provide an example of an external mechanism?

Reply:
Good idea. I added a paragraph on the notion of a discovery mechanism in https://github.com/libp2p/specs/pull/323/commits/21a3a11463568170a4ccef684857be17b45f7578. Let me know what you think.

Reply:
Resolving here. Please feel free to un-resolve.